In [1]:
# Initial the imports

import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from imblearn.over_sampling import RandomOverSampler
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report
from imblearn.metrics import classification_report_imbalanced

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Load the CSV file  into a Pandas DataFrame
lending_data_df=pd.read_csv("Resources/lending_data.csv")

# Display the DataFrame
display(lending_data_df.head())
display(lending_data_df.tail())                         

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

loan_size  interest_rate  borrower_income  debt_to_income  \
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Split the data into labels/target and features two sets

# Create the y variable, the labels
y = lending_data_df["loan_status"]

# Create the X variable, the features
X = lending_data_df.drop(columns="loan_status")

In [4]:
# Display the y variable Series
y.iloc[:15].values

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [5]:
# Display the X variable DataFrame
display(X.head(10))

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   
5    10100.0          7.438            50600        0.407115                4   
6    10300.0          7.490            51100        0.412916                4   
7     8800.0          6.857            45100        0.334812                3   
8     9300.0          7.096            47400        0.367089                3   
9     9700.0          7.248            48800        0.385246                4   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  
5                 1       20600  
6                 1       21100  
7                 0       15100  
8                 0       17400  
9                 0       18800

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target/label variables
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:

# Devide the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)
X_train.shape



(58152, 7)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
risk_classifier = LogisticRegression(max_iter=150,
                                random_state=1)
risk_classifier

# Fit the model using training data
risk_classifier.fit(X_train,y_train)

LogisticRegression(max_iter=150, random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
risk_predictions = risk_classifier.predict(X_test)
outcomes=pd.DataFrame({"Prediction": risk_predictions, "Actual": y_test}).reset_index(drop=True)
outcomes.head(10)

Prediction  Actual
0           0       0
1           0       1
2           0       0
3           0       0
4           0       0
5           0       0
6           0       0
7           0       0
8           0       0
9           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Calculate and print the balanced_accuracy score of the model
bal_accu_score = balanced_accuracy_score(y_test, risk_predictions)
print(f"Balanced Accuracy Score: {bal_accu_score}")

Balanced Accuracy Score: 0.9442676901753825


In [11]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, risk_predictions)

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [12]:
# Generate and print the classification report for the model
testing_report = classification_report(y_test, risk_predictions)
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** With the balanced accuracy score of 0.9442, the logistic regression model performs well detecting the credit risk of loan as a whole, especially for "0" healthy loan part with fi-score 1.00. 
  But for "1" high-risk loan part,  the comparatively lower score of precision and recall, means it might miss some instances which belong to this class; and only 625 support samples are chosen, indicating imbalanced sampled data.
  Overall, we suggest that resampling the dataset is necessary to be more accurate and convinsible. 

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [13]:
# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_ros, y_ros = ros.fit_resample(X_train, y_train)

In [14]:
# Ensure the resampled labels have an equal number of data points
y_ros.value_counts()


0    56277
1    56277
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [15]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model

risk_classifier_ros = LogisticRegression(max_iter=150,
                                random_state=1)


# Fit the model using training data
# Fit the model using the resampled training data
risk_classifier_ros.fit(X_ros,y_ros)

# Make a prediction using the testing data
risk_predictions_ros = risk_classifier_ros.predict(X_test)
outcomes_ros=pd.DataFrame({"Prediction": risk_predictions_ros, "Actual": y_test}).reset_index(drop=True)
outcomes_ros.head(10)

Prediction  Actual
0           0       0
1           1       1
2           0       0
3           0       0
4           0       0
5           0       0
6           0       0
7           0       0
8           0       0
9           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Calculate and print the balanced_accuracy score for the model using resampled data
bal_accu_score_ros = balanced_accuracy_score(y_test, risk_predictions_ros)
print(f"Balanced Accuracy Score: {bal_accu_score_ros}")

Balanced Accuracy Score: 0.9959744975744975


In [17]:
# Generate a confusion matrix for the model using resampled data
confusion_matrix(y_test, risk_predictions_ros)

array([[18668,    91],
       [    2,   623]], dtype=int64)

In [18]:
# Generate and print the classification report for the model using resampled data
testing_report = classification_report(y_test, risk_predictions_ros)
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      1.00      0.93       625

    accuracy                           1.00     19384
   macro avg       0.94      1.00      0.96     19384
weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model, fit with oversampled data, improves the performance obviously.For label "1" (high-risk loan),  recall score( 0.89 vs 1.00) and overall f1-score(0.88 vs 0.93) are increased contrastly, meaning the model is able to identify most instances correctly and classify less false positive samples.